x/ccv/provider/keeper/relay.go

```diff
@@ -153,7 +153,7 @@ func (k Keeper) sendValidatorUpdates(ctx sdk.Context) {
 			// check whether there is an established CCV channel to this consumer chain
 			if channelID, found := k.GetChainToChannel(ctx, chainID); found {
 				// send this validator set change packet data to the consumer chain
-				err := utils.SendIBCPacket(
+				expiredClient, err := utils.SendIBCPacket(
```

Comment:
Maybe we can refactor the whole `sendValidatorUpdates` keeper method and use the refactored approach in every other places we use `utils.SendIBCPacket`.

`SendIBCPacket` has more validation/preparation logic than it does send logic.

Something like this could help simplify error handling and client expiration tracking (sorry for the code dump):

```golang
// x/ccv/utils/utils.go
// expireClient is a good idea, but it can be handled without touching the actual send method and the logic can be simplified.
// 
// Main idea is to split SendIBCPacket into 2 parts -> "prep work" that checks channels, capabilities and sequences & the actual send
// - any error in "prep" should result in panic (e.g missing capability should panic, missing sequence should panic)
// - the send step no longer needs to be its own function, it can be done with channelKeeper.SendPacket(ctx, capability, packetData)

// SendIBCPacket sends an IBC packet with packetData
// over the source channelID and portID
func SendIBCPacket(
	ctx sdk.Context,
	channelKeeper ccv.ChannelKeeper,
	channelCap *types.Capability
	packetData *channeltypes.Packet
) error {
	// send packet over IBC channel
	return channelKeeper.SendPacket(ctx, channelCap, packetData)
}

// PrepareIBCPacket checks that all conditions for sending a packet are fullfilled and returns a prepared packet.
func PrepareIBCPacket(
	ctx sdk.Context,
	scopedKeeper ccv.ScopedKeeper,
	channelKeeper ccv.ChannelKeeper,
	channelID string,
	portID string,
	packetData []byte,
	timeoutPeriod time.Duration) (*channeltypes.Packet, error) {
	channel, ok := channelKeeper.GetChannel(ctx, portID, channelID)
	if !ok {
		return nil, sdkerrors.Wrapf(channeltypes.ErrChannelNotFound, "channel not found for channel ID: %s", channelID)
	}
	channelCap, ok := scopedKeeper.GetCapability(ctx, host.ChannelCapabilityPath(portID, channelID))
	if !ok {
		return nil, sdkerrors.Wrap(channeltypes.ErrChannelCapabilityNotFound, "module does not own channel capability")
	}

	// get the next sequence
	sequence, found := channelKeeper.GetNextSequenceSend(ctx, portID, channelID)
	if !found {
		return nil, sdkerrors.Wrapf(
			channeltypes.ErrSequenceSendNotFound,
			"source port: %s, source channel: %s", portID, channelID,
		)
	}
	packet := channeltypes.NewPacket(
		packetData, sequence,
		portID, channelID,
		channel.Counterparty.PortId, channel.Counterparty.ChannelId,
		clienttypes.Height{}, uint64(ctx.BlockTime().Add(timeoutPeriod).UnixNano()),
	)

	return &packet, nil
}

// x/ccv/provider/keeper/relay.go

// SendValidatorUpdates sends latest validator updates to every registered consumer chain
func (k Keeper) sendValidatorUpdates(ctx sdk.Context) {
	k.IterateConsumerChains(ctx, func(ctx sdk.Context, chainID, clientID string) (stop bool) {
		// check whether there is an established CCV channel to this consumer chain
		channelID, found := k.GetChainToChannel(ctx, chainID)
		if found {
			// Send pending VSC packets to consumer chain
			k.SendPendingVSCPackets(ctx, chainID, channelID)
		}


		valUpdateID := k.GetValidatorSetUpdateId(ctx) 	        // get current ValidatorSetUpdateId
		valUpdates := k.stakingKeeper.GetValidatorUpdates(ctx)  // get the validator updates from the staking module
		// check whether there are changes in the validator set;
		// note that this also entails unbonding operations
		// w/o changes in the voting power of the validators in the validator set
		unbondingOps, _ := k.GetUnbondingOpsFromIndex(ctx, chainID, valUpdateID)
		if len(valUpdates) > 0 || len(unbondingOps) > 0 {
			packetData := ccv.NewValidatorSetChangePacketData(valUpdates, valUpdateID, k.ConsumeSlashAcks(ctx, chainID))

			if !found {
				// store the packet data to be sent once the CCV channel is established
				k.AppendPendingVSC(ctx, chainID, []ccv.ValidatorSetChangePacketData{packetData})
			}
                         
                         // we only care about the packet if there's no error
			packet, err := PrepareIBCPacket(
				ctx,
				k.scopedKeeper,
				k.channelKeeper,
				channelID,          // source channel id
				ccv.ProviderPortID, // source port id
				packetData.GetBytes(),
				k.GetParams(ctx).CcvTimeoutPeriod,
			)

			// REVIEWER NOTE:
			// the returned errs are
			// - channeltypes.ErrChannelNotFound
			// - channeltypes.ErrChannelCapabilityNotFound,
			// - channeltypes.ErrSequenceSendNotFound
			// I'm not sure all of them should panic
			if err != nil {
				// something went wrong when sending the packet
				panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
			}

			// this can also be utils.SentIBCPacket(...)
			err = channelKeeper.SendPacket(ctx, channelCap, packetData)
			if err != nil && clienttypes.ErrClientNotActive.Is(err) {
				// IBC client expired:
				// store the packet data to be sent once the client is upgraded
				k.AppendPendingVSC(ctx, chainID, []ccv.ValidatorSetChangePacketData{packetData})
			} else if err != nil {
				// something went wrong when sending the packet
				panic(fmt.Errorf("packet could not be sent over IBC: %w", err))
			} else {
				// successful send:
				// set the VSC send timestamp for this packet
				k.SetVscSendTimestamp(ctx, chainID, packetData.ValsetUpdateId, ctx.BlockTime())
			}	
		}
		return true // do not stop the iteration
	})
	k.IncrementValidatorSetUpdateId(ctx)
}

```